File: workflow-engine/src/routes/monitoring.py
"""
Workflow Engine 监控路由
提供工作流引擎监控和统计数据
"""
from fastapi import APIRouter, Request
from datetime import datetime
from typing import Dict, Any, List, Optional
import time
import psutil

from shared_libs.luminaos_common.schemas.monitoring_schemas import (
    ServiceMonitoringData,
    ServiceHealth,
    ServiceStatus,
    APIStatistics,
    WorkflowExecutionStats,
    SystemResourceUsage,
    Metric,
    MetricType,
    MetricPoint,
    Alert,
    AlertLevel
)
from ..workflows import workflow_manager
from ..config import settings

router = APIRouter()

# 内存中的统计数据
_api_stats: Dict[str, Dict[str, Any]] = {}
_workflow_execution_stats: Dict[str, Dict[str, Any]] = {}
_start_time = time.time()


def _get_api_key(method: str, path: str) -> str:
    """生成API键"""
    return f"{method}:{path}"


def _update_api_stats(method: str, path: str, status_code: int, response_time: float):
    """更新API统计"""
    key = _get_api_key(method, path)
    
    if key not in _api_stats:
        _api_stats[key] = {
            "endpoint": path,
            "method": method,
            "total_requests": 0,
            "success_count": 0,
            "error_count": 0,
            "response_times": []
        }
    
    stats = _api_stats[key]
    stats["total_requests"] += 1
    
    if 200 <= status_code < 400:
        stats["success_count"] += 1
    else:
        stats["error_count"] += 1
    
    stats["response_times"].append(response_time)
    
    if len(stats["response_times"]) > 1000:
        stats["response_times"] = stats["response_times"][-1000:]


def _update_workflow_execution_stats(
    workflow_id: str,
    workflow_name: str,
    success: bool,
    execution_time: float
):
    """更新工作流执行统计"""
    if workflow_id not in _workflow_execution_stats:
        _workflow_execution_stats[workflow_id] = {
            "workflow_id": workflow_id,
            "workflow_name": workflow_name,
            "total_executions": 0,
            "success_count": 0,
            "failed_count": 0,
            "running_count": 0,
            "execution_times": [],
            "total_execution_time": 0.0
        }
    
    stats = _workflow_execution_stats[workflow_id]
    stats["total_executions"] += 1
    
    if success:
        stats["success_count"] += 1
    else:
        stats["failed_count"] += 1
    
    stats["execution_times"].append(execution_time)
    stats["total_execution_time"] += execution_time
    
    if len(stats["execution_times"]) > 1000:
        stats["execution_times"] = stats["execution_times"][-1000:]


def _calculate_percentiles(values: List[float], percentile: float) -> float:
    """计算百分位数"""
    if not values:
        return 0.0
    sorted_values = sorted(values)
    index = int(len(sorted_values) * percentile / 100)
    return sorted_values[min(index, len(sorted_values) - 1)]


@router.get(
    "/monitoring",
    response_model=ServiceMonitoringData,
    summary="获取服务监控数据",
    description="返回Workflow Engine的监控数据，包括健康状态、工作流执行统计、API统计等",
    tags=["Monitoring"]
)
async def get_monitoring_data() -> ServiceMonitoringData:
    """
    获取服务监控数据
    """
    # 计算运行时间
    uptime = time.time() - _start_time
    
    # 获取工作流统计
    try:
        workflows = await workflow_manager.list_workflows(page=1, page_size=1000)
        total_workflows = len(workflows)
    except Exception:
        total_workflows = 0
    
    # 健康状态
    health = ServiceHealth(
        service_name="workflow-engine",
        status=ServiceStatus.HEALTHY,
        version="1.0.0",
        uptime=uptime,
        last_check=datetime.now(),
        checks={
            "total_workflows": total_workflows,
            "workflow_manager_initialized": True
        }
    )
    
    # API统计
    api_statistics = []
    for key, stats in _api_stats.items():
        response_times = stats.get("response_times", [])
        if response_times:
            avg_time = sum(response_times) / len(response_times)
            min_time = min(response_times)
            max_time = max(response_times)
            p95_time = _calculate_percentiles(response_times, 95)
            p99_time = _calculate_percentiles(response_times, 99)
        else:
            avg_time = min_time = max_time = p95_time = p99_time = 0.0
        
        api_statistics.append(APIStatistics(
            endpoint=stats["endpoint"],
            method=stats["method"],
            total_requests=stats["total_requests"],
            success_count=stats["success_count"],
            error_count=stats["error_count"],
            avg_response_time=avg_time * 1000,
            min_response_time=min_time * 1000,
            max_response_time=max_time * 1000,
            p95_response_time=p95_time * 1000,
            p99_response_time=p99_time * 1000
        ))
    
    # 工作流执行统计
    workflow_stats = []
    for workflow_id, stats in _workflow_execution_stats.items():
        execution_times = stats.get("execution_times", [])
        avg_execution_time = sum(execution_times) / len(execution_times) if execution_times else 0.0
        
        workflow_stats.append(WorkflowExecutionStats(
            workflow_id=stats["workflow_id"],
            workflow_name=stats["workflow_name"],
            total_executions=stats["total_executions"],
            success_count=stats["success_count"],
            failed_count=stats["failed_count"],
            running_count=stats["running_count"],
            avg_execution_time=avg_execution_time,
            total_execution_time=stats["total_execution_time"]
        ))
    
    # 系统资源使用
    try:
        cpu_percent = psutil.cpu_percent(interval=0.1)
        memory = psutil.virtual_memory()
        disk = psutil.disk_usage('/')
        network = psutil.net_io_counters()
        
        resource_usage = SystemResourceUsage(
            cpu_percent=cpu_percent,
            memory_used=memory.used,
            memory_total=memory.total,
            memory_percent=memory.percent,
            disk_used=disk.used,
            disk_total=disk.total,
            disk_percent=(disk.used / disk.total * 100) if disk.total > 0 else 0,
            network_bytes_sent=network.bytes_sent,
            network_bytes_recv=network.bytes_recv
        )
    except Exception:
        resource_usage = SystemResourceUsage()
    
    # 指标
    metrics = [
        Metric(
            name="workflow_count",
            metric_type=MetricType.GAUGE,
            description="工作流总数",
            unit="count",
            points=[
                MetricPoint(
                    timestamp=datetime.now(),
                    value=float(total_workflows),
                    labels={"service": "workflow-engine"}
                )
            ]
        ),
        Metric(
            name="total_workflow_executions",
            metric_type=MetricType.COUNTER,
            description="总工作流执行次数",
            unit="count",
            points=[
                MetricPoint(
                    timestamp=datetime.now(),
                    value=float(sum(s["total_executions"] for s in _workflow_execution_stats.values())),
                    labels={"service": "workflow-engine"}
                )
            ]
        )
    ]
    
    # 告警
    alerts = []
    if resource_usage.cpu_percent > 80:
        alerts.append(Alert(
            id=f"cpu_high_{int(time.time())}",
            level=AlertLevel.WARNING,
            title="CPU使用率过高",
            message=f"CPU使用率达到 {resource_usage.cpu_percent:.1f}%",
            service="workflow-engine",
            created_at=datetime.now(),
            metadata={"cpu_percent": resource_usage.cpu_percent}
        ))
    
    if resource_usage.memory_percent > 80:
        alerts.append(Alert(
            id=f"memory_high_{int(time.time())}",
            level=AlertLevel.WARNING,
            title="内存使用率过高",
            message=f"内存使用率达到 {resource_usage.memory_percent:.1f}%",
            service="workflow-engine",
            created_at=datetime.now(),
            metadata={"memory_percent": resource_usage.memory_percent}
        ))
    
    # 注意：这里我们需要将 workflow_stats 转换为 API 可接受的格式
    # 由于 ServiceMonitoringData 没有 workflow_execution_stats 字段，我们将其放在 metadata 中
    return ServiceMonitoringData(
        service_name="workflow-engine",
        health=health,
        api_statistics=api_statistics,
        metrics=metrics,
        resource_usage=resource_usage,
        alerts=alerts,
        last_updated=datetime.now()
    )


@router.get(
    "/monitoring/workflow-stats",
    summary="获取工作流执行统计",
    description="返回工作流执行统计信息",
    tags=["Monitoring"]
)
async def get_workflow_execution_stats() -> Dict[str, Any]:
    """获取工作流执行统计"""
    workflow_stats = []
    for workflow_id, stats in _workflow_execution_stats.items():
        execution_times = stats.get("execution_times", [])
        avg_execution_time = sum(execution_times) / len(execution_times) if execution_times else 0.0
        
        workflow_stats.append({
            "workflow_id": stats["workflow_id"],
            "workflow_name": stats["workflow_name"],
            "total_executions": stats["total_executions"],
            "success_count": stats["success_count"],
            "failed_count": stats["failed_count"],
            "running_count": stats["running_count"],
            "avg_execution_time": avg_execution_time,
            "total_execution_time": stats["total_execution_time"]
        })
    
    return {
        "workflow_execution_stats": workflow_stats,
        "timestamp": datetime.now().isoformat()
    }


# 中间件函数
async def monitoring_middleware(request: Request, call_next):
    """监控中间件，记录请求统计"""
    start_time = time.time()
    response = await call_next(request)
    response_time = time.time() - start_time
    
    _update_api_stats(
        method=request.method,
        path=str(request.url.path),
        status_code=response.status_code,
        response_time=response_time
    )
    
    return response

File: web-ui/src/lib/api/monitoring.ts
/**
 * 监控API客户端
 */
import { mcpGatewayClient, apiGatewayClient } from './client';

export interface ServiceHealth {
  service_name: string;
  status: 'healthy' | 'unhealthy' | 'degraded' | 'down' | 'unknown';
  version: string;
  uptime: number;
  last_check: string;
  checks: Record<string, any>;
}

export interface APIStatistics {
  endpoint: string;
  method: string;
  total_requests: number;
  success_count: number;
  error_count: number;
  avg_response_time: number;
  min_response_time: number;
  max_response_time: number;
  p95_response_time: number;
  p99_response_time: number;
}

export interface SystemResourceUsage {
  cpu_percent: number;
  memory_used: number;
  memory_total: number;
  memory_percent: number;
  disk_used: number;
  disk_total: number;
  disk_percent: number;
  network_bytes_sent: number;
  network_bytes_recv: number;
}

export interface Alert {
  id: string;
  level: 'info' | 'warning' | 'error' | 'critical';
  title: string;
  message: string;
  service: string;
  created_at: string;
  resolved_at?: string;
  resolved: boolean;
  metadata: Record<string, any>;
}

export interface Metric {
  name: string;
  metric_type: string;
  description: string;
  unit?: string;
  points: Array<{
    timestamp: string;
    value: number;
    labels: Record<string, string>;
  }>;
}

export interface ServiceMonitoringData {
  service_name: string;
  health: ServiceHealth;
  api_statistics: APIStatistics[];
  metrics: Metric[];
  resource_usage?: SystemResourceUsage;
  alerts: Alert[];
  last_updated: string;
}

export interface MonitoringOverview {
  services: ServiceMonitoringData[];
  total_alerts: number;
  critical_alerts: number;
  system_resource?: SystemResourceUsage;
  overall_status: string;
  timestamp: string;
}

export interface WorkflowExecutionStats {
  workflow_id?: string;
  workflow_name?: string;
  total_executions: number;
  success_count: number;
  failed_count: number;
  running_count: number;
  avg_execution_time: number;
  total_execution_time: number;
}

export interface UserActivityStats {
  user_id?: string;
  username?: string;
  active_sessions: number;
  total_requests: number;
  last_activity?: string;
  login_count: number;
}

export interface LogEntry {
  timestamp: string;
  level: string;
  service: string;
  logger: string;
  message: string;
  context: Record<string, any>;
  traceback?: string;
}

export interface LogQueryParams {
  service?: string;
  level?: string;
  start_time?: string;
  end_time?: string;
  keyword?: string;
  limit?: number;
  offset?: number;
}

export interface LogQueryResponse {
  success: boolean;
  logs: LogEntry[];
  total: number;
  limit: number;
  offset: number;
  timestamp: string;
}

/**
 * 获取MCP Gateway监控数据
 */
export async function getMCPGatewayMonitoring(): Promise<ServiceMonitoringData> {
  try {
    return await mcpGatewayClient.get<ServiceMonitoringData>('/monitoring');
  } catch (error: any) {
    if (error?.statusCode === 404) {
      return mcpGatewayClient.get<ServiceMonitoringData>('/api/monitoring');
    }
    throw error;
  }
}

/**
 * 获取Workflow Engine监控数据
 */
export async function getWorkflowEngineMonitoring(): Promise<ServiceMonitoringData> {
  // 监控API在 /api/v1/monitoring，通过 API Gateway 访问
  return apiGatewayClient.get<ServiceMonitoringData>('/api/v1/monitoring');
}

/**
 * 获取Auth Service监控数据
 */
export async function getAuthServiceMonitoring(): Promise<ServiceMonitoringData> {
  const response = await fetch('/api/auth-monitoring');
  if (!response.ok) {
    throw new Error(`HTTP error! status: ${response.status}`);
  }
  return response.json();
}

/**
 * 获取所有服务监控数据
 */
export async function getAllServicesMonitoring(): Promise<MonitoringOverview> {
  const [mcpGateway, workflowEngine, authService] = await Promise.allSettled([
    getMCPGatewayMonitoring(),
    getWorkflowEngineMonitoring(),
    getAuthServiceMonitoring(),
  ]);

  const services: ServiceMonitoringData[] = [];

  if (mcpGateway.status === 'fulfilled') {
    services.push(mcpGateway.value);
  }
  if (workflowEngine.status === 'fulfilled') {
    services.push(workflowEngine.value);
  }
  if (authService.status === 'fulfilled') {
    services.push(authService.value);
  }

  const total_alerts = services.reduce((sum, s) => sum + s.alerts.length, 0);
  const critical_alerts = services.reduce(
    (sum, s) => sum + s.alerts.filter((a) => a.level === 'critical' || a.level === 'error').length,
    0
  );

  // 计算整体状态
  let overall_status = 'healthy';
  if (services.some((s) => s.health.status === 'down')) {
    overall_status = 'down';
  } else if (services.some((s) => s.health.status === 'unhealthy')) {
    overall_status = 'unhealthy';
  } else if (services.some((s) => s.health.status === 'degraded')) {
    overall_status = 'degraded';
  }

  return {
    services,
    total_alerts,
    critical_alerts,
    system_resource: services[0]?.resource_usage,
    overall_status,
    timestamp: new Date().toISOString(),
  };
}

/**
 * 获取工作流执行统计
 */
export async function getWorkflowExecutionStats(): Promise<{
  workflow_execution_stats: WorkflowExecutionStats[];
  timestamp: string;
}> {
  // 监控API在 /api/v1/monitoring，不在 /api/workflows 下
  // 所以需要直接通过 API Gateway 访问
  const API_GATEWAY_URL = process.env.NEXT_PUBLIC_API_GATEWAY_URL || 'http://127.0.0.1:8080';
  try {
    const response = await fetch(`${API_GATEWAY_URL}/api/v1/monitoring/workflow-stats`, {
      method: 'GET',
      headers: {
        'Content-Type': 'application/json',
      },
    });

    if (!response.ok) {
      if (response.status === 404) {
        // 柔性降级：如果接口不存在或返回404，则返回空数据，避免前端报错
        return { workflow_execution_stats: [], timestamp: new Date().toISOString() };
      }
      throw new Error(`HTTP error! status: ${response.status}`);
    }

    return response.json();
  } catch (e: any) {
    // 柔性降级：如果接口不存在或返回404，则返回空数据，避免前端报错
    if (e?.statusCode === 404 || e?.status === 404 || e?.message?.includes('404')) {
      return { workflow_execution_stats: [], timestamp: new Date().toISOString() };
    }
    throw e;
  }
}

/**
 * 获取用户活跃度统计
 */
export async function getUserActivityStats(): Promise<{
  user_activity_stats: UserActivityStats[];
  timestamp: string;
}> {
  try {
    const response = await fetch('/api/auth-monitoring/user-activity');
    if (!response.ok) {
      if (response.status === 404) {
        return { user_activity_stats: [], timestamp: new Date().toISOString() };
      }
      throw new Error(`HTTP error! status: ${response.status}`);
    }
    return response.json();
  } catch (e: any) {
    if (e?.statusCode === 404 || e?.status === 404 || e?.message?.includes('404')) {
      return { user_activity_stats: [], timestamp: new Date().toISOString() };
    }
    throw e;
  }
}

/**
 * 查询日志
 */
export async function queryLogs(params: LogQueryParams): Promise<LogQueryResponse> {
  const queryString = new URLSearchParams();
  if (params.service) queryString.append('service', params.service);
  if (params.level) queryString.append('level', params.level);
  if (params.start_time) queryString.append('start_time', params.start_time);
  if (params.end_time) queryString.append('end_time', params.end_time);
  if (params.keyword) queryString.append('keyword', params.keyword);
  if (params.limit) queryString.append('limit', params.limit.toString());
  if (params.offset) queryString.append('offset', params.offset.toString());

  // 这里应该有一个统一的日志查询端点，暂时返回空结果
  // 实际实现需要后端提供统一的日志查询API
  return {
    success: true,
    logs: [],
    total: 0,
    limit: params.limit || 100,
    offset: params.offset || 0,
    timestamp: new Date().toISOString(),
  };
}
